File: 2641-disconnect-path-in-a-binary-matrix-by-at-most-one-flip/disconnect-path-in-a-binary-matrix-by-at-most-one-flip.cpp
class Solution {
public:
    int n, m;
    bool dfs(int row, int col, vector<vector<int>>& grid)
    {
        if(row >= n || col >= m || !grid[row][col])
        return false;

        if(row == n - 1 && col == m - 1)
        return true;

        // marking the current cell as visited
        grid[row][col] = 0;

        return dfs(row, col + 1, grid) || dfs(row +1, col, grid);
    }

    bool isPossibleToCutPath(vector<vector<int>>& grid) 
    {
        n = grid.size();
        m = grid[0].size();

        // we can't even reach the last cell already then no need of flip
        // this dfs will mark one of the path taken from src to dest as '0's
        if(!dfs(0, 0, grid))
        return true;

        // mark the src cell as unvisited.
        grid[0][0] = 1;

        if(dfs(0, 0, grid))
        return false;

        return true;
    }
};

File: 2641-disconnect-path-in-a-binary-matrix-by-at-most-one-flip/disconnect-path-in-a-binary-matrix-by-at-most-one-flip.java
class Solution {

    private int n, m;
    private boolean canReachBottomRightCell(int row, int col, int[][] grid){
        if(row < 0 || row >= grid.length || col < 0 || col >= grid[0].length || grid[row][col] == 0){
            return false;
        }

        if(row == n - 1 && col == m - 1)
            return true;

        // Mark the current cell as visited
        grid[row][col] = 0;

        return canReachBottomRightCell(row, col + 1, grid) || canReachBottomRightCell(row + 1, col, grid);
    }

    public boolean isPossibleToCutPath(int[][] grid) {
        n = grid.length;
        m = grid[0].length;

        // matrix is disconnected if there exists no path already
        if(!canReachBottomRightCell(0, 0, grid)){
            return true;
        }

        // Restore the (0,0) cell to 1
        grid[0][0] = 1;

        // if you reach the bottom right cell
        return !canReachBottomRightCell(0, 0, grid);
    }
}